subgraphs/aave-v2/src/mappings/utilFunctions.ts

```diff
@@ -0,0 +1,559 @@
+  }
+  const currentPriceUSD = getAssetPriceInUSDC(Address.fromString(market.id));
+  market.outputTokenPriceUSD = currentPriceUSD;
+  market.inputTokenPricesUSD = [currentPriceUSD];
+  market.save();
+  return market as Market;
+}
+
+export function getCurrentRewardEmissions(market: Market): BigInt[] {
+  const rewardEmissions = market.rewardTokenEmissionsAmount;
+  // Attempt to get the incentives controller contract
+  const incentivesController = initIncentivesController(market);
+  // From the incentives controller contract, get pull the 'assets' values from the market aToken, sToken, and vToken
+  const incentivesControllerContract = IncentivesControllerContract.bind(incentivesController);
+  if (!incentivesControllerContract.try_assets(Address.fromString(market.outputToken)).reverted) {
+    log.info('REWARD EMISSIONS ' + market.outputToken + ' ' + incentivesControllerContract.assets(Address.fromString(market.outputToken)).value0.toString(), []);
+    const assetDataSupply = incentivesControllerContract.try_assets(Address.fromString(market.outputToken)).value.value0;
+    const assetDataBorrowStable = incentivesControllerContract.try_assets(Address.fromString(market.sToken)).value.value0;
+    const assetDataBorrowVariable = incentivesControllerContract.try_assets(Address.fromString(market.vToken)).value.value0;
+    // Get the emissions per day for the aToken rewards for deposits
+    if (!assetDataSupply.equals(BIGINT_ZERO)) {
+      rewardEmissions[0] = emissionsPerDay(assetDataSupply);
+    } else {
+      rewardEmissions[0] = BIGINT_ZERO;
+    }
+    // Get the emissions per second for both the sToken and vToken rewards, average them and get the daily emissions for borrows
+    const borrowRewardsAvgRate = (assetDataBorrowStable.plus(assetDataBorrowVariable)).div(BIGINT_TWO);
+    log.info('BORROW AVG RATE: ' + assetDataBorrowStable.toString() + ' + ' + assetDataBorrowVariable.toString() + ' = ' + (assetDataBorrowStable.plus(assetDataBorrowVariable)).toString() + ' /2 = ' + borrowRewardsAvgRate.toString(), [])
+    if (!borrowRewardsAvgRate.equals(BIGINT_ZERO)) {
+      rewardEmissions[1] = emissionsPerDay(borrowRewardsAvgRate);
+    } else {
+      rewardEmissions[1] = BIGINT_ZERO;
+    }
+  } else {
+    log.info('COULD NOT GET REWARD EMISSIONS FROM INC CONT ON MARKET: ' + market.id, []);
+    // The rewardEmissions array returned is already defaulted to zero vals
+  }
+  return rewardEmissions;
+}
+
+export function getOutputTokenSupply(event: ethereum.Event): void {
+  const outputTokenAddr = event.address;
+  const aTokenInstance = AToken.bind(outputTokenAddr);
+  const tryTokenSupply = aTokenInstance.try_totalSupply();
+  if (!tryTokenSupply.reverted) {
+    log.info('OUTPUT TOKEN NOT REVERTED ' + outputTokenAddr.toHexString() + ' ' + tryTokenSupply.value.toString(), []);
+    const market = Market.load(aTokenInstance.UNDERLYING_ASSET_ADDRESS().toHexString());
```

Comment:
You should save `UNDERLYING_ASSET_ADDRESS` so you don't have to query it every time. Bear in mind each contract call takes 200-500ms.

Reply:
Added it to the Token entity. If the Token being created is an aToken, the underlying asset is added to the entity. Otherwise, it is saved as an empty string. Now it can be pulled just from the Token instance of the aToken rather than making a contract call .

Reply:
Added it to the Token entity. If the Token being created is an aToken, the underlying asset is added to the entity. Otherwise, it is saved as an empty string. Now it can be pulled just from the Token instance of the aToken rather than making a contract call .